staging/src/k8s.io/client-go/tools/cache/reflector.go

```
@@ -483,6 +484,17 @@ func (r *Reflector) watch(ctx context.Context, w watch.Interface, resyncerrc cha
 	var err error
 	retry := NewRetryWithDeadline(r.MaxInternalErrorRetryDuration, time.Minute, apierrors.IsInternalError, r.clock)
 
+	defer func() {
+		if r := recover(); r != nil {
+			buf := make([]byte, 1<<16)
+			stackSize := golangruntime.Stack(buf, false)
+			stackTrace := string(buf[:stackSize])
+			klog.Errorf("Panic occurred in Reflector watch operation: %v\nStack Trace:\n%s", r, stackTrace)
```

No direct klog calls, please. Use the `logger` instance instead.

What is the advantage of this custom code over [runtime.HandleCrashWithContext](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/runtime#HandleCrashWithContext)?

---

Thanks for the suggestion! I used it in `watch` but not `list`. Here are my reasons:

1. [runtime.HandleCrashWithContext](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/runtime#HandleCrashWithContext) by default re-panics after calling the handlers. That doesn't work here because the panic happens inside a goroutine in `list()`, and we need to pass the panic back to the main goroutine via `panicCh`. Re-panicking inside the child goroutine would bypass that and crash the process.

2. Setting `ReallyCrash = false` would avoid the re-panic, but it's a global variable and not thread-safe. Toggling it can trigger data races in tests or other concurrent code.

3. So instead, I used a local `recover()` block that mimics `HandleCrashWithContext` but without the re-panic. It still calls the registered panic handlers (so we get the logs and stack trace), and then sends the panic over the channel for the caller to handle.

4. If we think this pattern is useful more broadly, I’d be happy to propose a patch to `staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go` to support it more cleanly.

---

nvm, just after I wrote these, I found I can make it cleaner by using

```
	defer func() {
		if r := recover(); r != nil {
			panicCh <- r
		}
	}()

	defer utilruntime.HandleCrashWithContext(ctx)
```

---

> sends the panic over the channel for the caller to handle

Who needs this? What is the caller supposed to do with the panic?

We seem to have a policy in Kubernetes which is implemented by `k8s.io/apimachinery/pkg/util/runtime`: "panics should not occur and are fatal". If you feel that this  code here needs a different policy, then I would like to see confirmation from SIG Arch on that.

I agree that *for testing purposes* it makes sense to apply a different policy, and I agree that the global variable in the runtime package is sub-optimal. If that is what you are trying to fix, then we could look for simpler alternatives. For example, a "don't exit" flag could be set in the context and get checked by the package first before falling back to the global variable.

---

Looking at this closer, I understand why `panicCh` exists: `list` spawns a goroutine and any panic occurring there is meant to be seen as a panic in `list` itself.

Your latest incarnation doesn't achieve that. When calling `utilruntime.HandleCrashWithContext(ctx)` inside the goroutine, the entire process is getting torn down (default behavior outside of tests), without ever writing anything to `panicCh`.

Perhaps that *is* the right approach. But then the `panicCh` is redundant and should be removed.

---

Thanks for the detailed feedback! Just to clarify, my latest version **does not let the process crash** from `HandleCrashWithContext(ctx)`'s internal re-panic.

Here's why: there's an outer `defer recover()` that wraps the goroutine, placed *before* the `defer utilruntime.HandleCrashWithContext(ctx)` line. So even though `HandleCrashWithContext` re-panics by default, that panic is caught by the outer `recover()` and forwarded to `panicCh`, which then gets handled in the `list()` caller goroutine.

So both of these are satisfied:
1. The full stacktrace from `utilruntime.PanicHandlers` gets logged (this is required for cases requiring `expectLogSub: []string{"Observed a panic", "simulated panic in ListFunc", "stacktrace"}` to pass).
2. The panic doesn’t bring down the process, even when `ReallyCrash` is `true`, or my tests wouldn't have passed.

That’s why `panicCh` is necessary here. You are right that it ensures that the panic originating inside the `list()` goroutine is treated as a panic from `list()` itself. The latest implementation preserves the control flow without making major changes.

Regarding the long-term direction: I agree that having a flag (e.g. in context) instead of using `ReallyCrash` would be a cleaner general-purpose mechanism. But after making this double-defer pattern work cleanly, I feel it might be enough to document this use case (or offer it as an opt-in helper in `runtime.go`). Would you like to share your opinion on this?

All tests are passing now, including the ones that require both log capture *and* controlled panic propagation. Let me know if there are any other concerns!

---

Just to clarify, I didn’t just drop `defer utilruntime.HandleCrashWithContext(ctx)` in and assume it would work.

At one point, I *replaced* the original `recover()` block entirely with `HandleCrashWithContext`. But when I realized that its internal re-panic wasn’t being caught, and was crashing the process, I brought the outer `recover()` back, specifically to catch that re-panic and forward it via `panicCh`.

So although the outer `defer recover()` might look unchanged in the file diff, its purpose has completely shifted:  
- Previously, it was doing all the panic handling.  
- Now, `HandleCrashWithContext` handles logging, and the outer `recover()` catches the re-panic for safe propagation.

I just wanted to highlight this since the change might not be obvious from the diff alone.

---

> I'm also personally lean toward refactoring this function in a synchronous style.

Big :+1:  for that.

Some arguments why the current approach is flawed or at least questionable:
- If [pager.ListWithAlloc](https://github.com/kubernetes/kubernetes/blob/88dfcb225d41326113990e87b11137641c121a32/staging/src/k8s.io/client-go/tools/cache/reflector.go#L611C32-L611C51) panics after `list` has returned because of context cancellation, then the panic is silently swallowed and not handled.
- Contextual logging in pager.ListWithAlloc does not work - or perhaps it works because the pager injects back the [actual context](https://github.com/kubernetes/kubernetes/blob/88dfcb225d41326113990e87b11137641c121a32/staging/src/k8s.io/client-go/tools/cache/reflector.go#L586), but it's convoluted.

---

Hi @pohly Thanks again for your earlier feedback! I’ve updated the PR based on our previous discussion.

Here's the changelist:
- `Reflector.list` is now fully synchronous, panic handling is simplified.
- Used `HandleCrashWithContext(ctx)` in both `list` and `watch`.
- Updated `TestUntilWithSync` to use `ListWithContextFunc`, so it properly respects timeouts.
- Legacy deprecated `ListFunc` in `TestUntilWithSync` cases that block (`select {}`) now hang in sync mode. I added comments and excluded those from test runs to avoid timeouts. 
- Added `TestReflectorListWatchPanicLoggingWithContext` to test panics in context-aware list/watch functions.

Let me know if you'd suggest any further changes or if this direction looks good to you!

---

Hey @pohly, just wanted to clarify a bit more around potential blocking concerns after switching to a synchronous implementation:

Previously, the async `list` implementation would silently "hide" user errors like an indefinitely blocking `ListFunc` by exiting the main goroutine early on context cancellation, while the underlying goroutine remained stuck. This covered up a misuse pattern that could lead to runtime issues such as goroutine leaks.

The new synchronous version surfaces this problem more explicitly. If a user mistakenly provides a blocking (non-context-aware) `ListFunc`, the call will now block indefinitely, and if such legacy test cases are included, the Go test framework will eventually timeout and fail.

Since `ListFunc` is deprecated and the context-aware versions are strongly encouraged, I think making this kind of misuse more visible is actually beneficial. I didn't to include an explicit test for this behavior, as it would require letting the test hang until timeout, which is not desirable. Instead, I've included comments in the test suite to document the legacy behavior and explain why it's no longer tested.

Let me know what you think when you have a chance!

---

NewReflectorWithOptions still takes a ListerWatcher that does not include context in the signature.

It *checks* if there's a context-aware signature, but if there's not, it wraps it in an adapter that discards the context:
```
func (l listerWrapper) ListWithContext(ctx context.Context, options metav1.ListOptions) (runtime.Object, error) {
	return l.parent.List(options)
}
```

This PR is now *relying* on the lister used by the pager to honor the context and cancel and return immediately when the context is canceled, in order to make the documented behavior of Run() work properly, and ListAndWatch return when it's stop channel is closed.

```
// RunWithContext repeatedly uses the reflector's ListAndWatch to fetch all the
// objects and subsequent deltas.
// Run will exit when the context is canceled.
func (r *Reflector) RunWithContext(ctx context.Context) {
```

If we're going to make context cancellation of the listerWatcher load-bearing, we probably have to move the background list + foreground early return on ctx.cancellation into that adapter.

---

To be clear, I'm in favor of this PR as far as it goes, plumbing ctx to the lister so we don't hang a background list. I just think we need to make our ctx adapter more correct if we're relying on it to interrupt the list for a context cancel OR remove the possibility of listerwatchers that don't have context

(just a quick look in-tree shows ~9 places we're constructing cache.ListWatch with only ListFunc/WatchFunc methods, not the context equivalents ... looks like we haven't really finished the work to make lister watcher context aware ... so relying on it in this PR isn't super solid yet)

---

/hold

---

actually, it looks like we have multiple implementations of ListWithContext now that just drop the context:

```
// List a set of apiserver resources
func (lw *ListWatch) ListWithContext(ctx context.Context, options metav1.ListOptions) (runtime.Object, error) {
	// ListWatch is used in Reflector, which already supports pagination.
	// Don't paginate here to avoid duplication.
	if lw.ListWithContextFunc != nil {
		return lw.ListWithContextFunc(ctx, options)
	}
	return lw.ListFunc(options)
}
```

```
func (l listerWrapper) ListWithContext(ctx context.Context, options metav1.ListOptions) (runtime.Object, error) {
	return l.parent.List(options)
}
```

if we're relying on ctx cancellation here, we have to make those actually cancel on context cancel

---

We can fix all in-tree code, but ListAndWatch is a public API. There's always the possibility that some client-go consumer only supplies the non-context-aware functions. Perhaps we have to really remove those before we can proceed with this PR?

---

I think the minimum bar to rely on context cancellation here is:
1. converting all in-tree uses to honor context
2. making the places we adapt context-unaware listers to context-aware signatures unblock and return on context cancellation (leaving the uncancellable lists running/hanging in the background, I guess, with the panic propagation workaround shifted to those adapters)

To help move more of the out-of-tree ecosystem to threading context cancellation, we have a spectrum of options:
1. Sweep the internet for things that use our concrete ListWatch struct with ListFunc/WatchFunc and file issues or open PRs to switch them to provide ListWithContextFunc / WatchWithContextFunc instead. This is beneficial even if we never do any subsequent steps.
2. Possibly, after we've done 1 and waited some period of time, remove the non-context ListFunc/WatchFunc fields from `ListWatch` (while keeping the methods to satisfy the interface)
3. Possibly, after we've done 2 and waited some period of time, collapse the ListerWatcher and ListerWatcherWithContext interfaces.